src/Sulu/Bundle/TagBundle/Tag/TagManager.php

```diff
@@ -143,6 +155,54 @@ public function save($data, $id = null)
         }
     }
 
+    public function restore(int $id, array $data): TagInterface
+    {
+        $name = $data['name'];
+
+        /** @var TagInterface|null $existingTag */
+        $existingTag = $this->tagRepository->findTagById($id);
+        if (null !== $existingTag) {
+            throw new TagAlreadyExistsException($existingTag->getName());
+        }
+
+        /** @var TagInterface|null $existingTag */
+        $existingTag = $this->tagRepository->findTagByName($name);
+        if (null !== $existingTag) {
+            throw new TagAlreadyExistsException($existingTag->getName());
+        }
+
+        $tag = $this->tagRepository->createNew();
+        $tag->setName($name);
+
+        $tagClass = \get_class($tag);
+
+        $idReflProperty = new \ReflectionProperty($tagClass, 'id');
+        $idReflProperty->setAccessible(true);
+        $idReflProperty->setValue($tag, $id);
+
+        /** @var ClassMetadataInfo<TagInterface> $metadata */
+        $metadata = $this->em->getClassMetaData($tagClass);
+
+        $oldIdGeneratorType = $metadata->generatorType;
+        $oldIdGenerator = $metadata->idGenerator;
+
+        $metadata->setIdGeneratorType(ClassMetadata::GENERATOR_TYPE_NONE);
+        $metadata->setIdGenerator(new AssignedGenerator());
+
+        $this->em->persist($tag);
+
+        $this->domainEventCollector->collect(
+            new TagRestoredEvent($tag, $data)
+        );
+
+        $this->em->flush();
+
+        $metadata->setIdGeneratorType($oldIdGeneratorType);
+        $metadata->setIdGenerator($oldIdGenerator);
```

Comment:
I'm not sure, if it's a good idea to set the id like this ... but is there a better way?